app/feature/feature-travel-certificate/src/main/kotlin/com/hedvig/android/feature/travelcertificate/ui/generate_when/TravelCertificateDateInputViewModel.kt

```diff
@@ -0,0 +1,270 @@
+  @Composable
+  override fun MoleculePresenterScope<TravelCertificateDateInputEvent>.present(
+    lastState: TravelCertificateDateInputUiState,
+  ): TravelCertificateDateInputUiState {
+    var loadIteration by remember { mutableIntStateOf(0) }
+
+    var generateIteration by remember { mutableIntStateOf(0) }
+
+    var screenContent by remember { mutableStateOf<DateInputScreenContent>(DateInputScreenContent.Loading) }
+
+    var travelDate by remember {
+      mutableStateOf(
+        Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault()).date,
+      )
+    }
+
+    var contractIdFromBackend by remember { mutableStateOf<String?>(null) }
+
+    var email by remember { mutableStateOf<String?>(null) }
+
+    var primaryInput by remember {
+      mutableStateOf<TravelCertificatePrimaryInput?>(null)
+    }
+
+    var isInputValid by remember { mutableStateOf<Boolean?>(null) }
+
+    var hasCoEnsured by remember { mutableStateOf(false) }
+
+    CollectEvents { event ->
+      when (event) {
+        TravelCertificateDateInputEvent.RetryLoadData -> {
+          loadIteration++
+        }
+
+        is TravelCertificateDateInputEvent.ChangeEmailInput -> {
+          email = event.email
+        }
+
+        is TravelCertificateDateInputEvent.ChangeDataInput -> {
+          travelDate = event.localDate
+        }
+
+        TravelCertificateDateInputEvent.NullifyInputValidity -> {
+          isInputValid = null
+        }
+
+        TravelCertificateDateInputEvent.ValidateInputAndChooseDirection -> {
+          val currentEmail = email
+          val isCurrentInputValid =
+            currentEmail != null && currentEmail.length > 5 // todo: couldn't be shorter than this, right?
+          isInputValid = isCurrentInputValid
+          if (isCurrentInputValid) {
+            if (hasCoEnsured) {
+              val currentContractId = contractIdFromBackend
+              val travelCertificatePrimaryInput = if (currentEmail != null &&
+                currentContractId != null
+              ) {
+                TravelCertificatePrimaryInput(
+                  currentEmail,
+                  travelDate,
+                  currentContractId,
+                )
+              } else {
+                null
+              }
+              if (travelCertificatePrimaryInput == null) {
+                logcat(LogPriority.INFO) {
+                  "TravelCertificateDateInputPresenter: currentEmail or currentContractId are null when they shouldn't be"
+                }
+              }
+              primaryInput = travelCertificatePrimaryInput
+            } else {
+              generateIteration++ //
+            }
+          } // todo: else isInputValid will be false will go to Success and show toast
+        }
+      }
+    }
+
+    LaunchedEffect(generateIteration) {
+      val currentContent = screenContent
+      val currentEmail = email
+      if (currentContent is DateInputScreenContent.Success && currentEmail != null) {
```

Comment:
we actually cannot go back to this screen if we go through generation, because we popUp this destination, we would go back to history screen. Do you think it is possible to be stuck in a loop in this case?